### ISSUE TITLE
How to load multiple models

### ISSUE THREAD

**Issue by CharlesShang** [tflearn/tflearn#293]
I want to load multiple trained models at a time, these models share some same tensors, but different weights. 
Lodaing them I got the `con_1/W already exists` errors.

One way to solve this is to add  `with tf.variable_scope('fr_en')` before each model at training time. ([checking this link](https://github.com/tensorflow/tensorflow/issues/3270) ) But I don't want to restrain these models, what should I do?

**Comment by aymericdamien** [comment]
Maybe you can encapsulate them with 'with tf.Graph().as_default():' to define a specific graph for each of them. Otherwise, try to use the argument 'restore=False' of layers if you do not want to initialize them.

**Comment by CharlesShang** [comment]
Thank you for your suggestions. But there are still some problems
I tried to save my models to graph def and weights using `tf.train.write_graph` and `tf.train.Saver(tf.all_variables())`, and restore them using tf.GraphDef() to restore the graph and Saver for the vars. It can be restored correctly. 

    def load_vgg_simple():
        # Load the network
        network = vgg16()
        model = tflearn.DNN(network)
        model.load('model_vgg/vgg16.tfl-500', weights_only=True)
        return model.session


    def load_vgg2_via_graph():

        with gfile.FastGFile('model_vgg/vgg_model2_graphdef.pb', 'rb') as f:
            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read()) # load graph
            graph = tf.Graph()
            graph.as_default()
            tf.import_graph_def(graph_def, name='') # set default
            # build a new sess by passing a graph
            with tf.Session(graph=graph) as persisted_sess:  
                network2 = vgg16_2()
                model2 = tflearn.DNN(network2)
                persisted_sess.run(tf.initialize_all_variables())
                saver = tf.train.Saver(tf.all_variables())
                saver.restore(persisted_sess, "model_vgg/vgg_model2.cp") # load vars
                return persisted_sess


I call the two function 

    sess1 = load_vgg_simple()
    sess2 = load_vgg2_via_graph()

    sess1.run()  # do sth correctly
    sess2.run()  # RuntimeError: Attempted to use a closed Session

How can I keep both the sessions alive during?

**Comment by aymericdamien** [comment]
in here `with tf.Session(graph=graph) as persisted_sess: `
You have to pass the session of the model object (`model.session`) instead of creating a new one.

**Comment by CharlesShang** [comment]
It works! Thank you, @aymericdamien 
can load multiple models by calling `load_vgg_simple(graph_def_path, model_path)` saveral times, without vars conflicts. But still need to find out the mechanism. 

def load_vgg_simple(graph_def_path, model_path):

    # Load the network
    with gfile.FastGFile(graph_def_path, 'rb') as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read()) # load graph
    graph = tf.Graph()
    graph.as_default()
    tf.import_graph_def(graph_def, name='')  # set default
    with tf.Session(graph=graph) as persisted_sess:
        network = vgg16(is_test=True)
        model = tflearn.DNN(network, session = persisted_sess)
        model.load(model_path, weights_only=True)
    return model.session

**Comment by eduard-cojocea** [comment]
I encountered the same problem and could not solve the problem (without retraining) with any solution i found on the internet. So what I did is load each model in two separate threads which communicate with the main thread. It is simple enough to write the code, you just have to be careful when you synchronize the threads. In my case each thread received the input for its problem and returned to the main thread the output. It works without any observable overhead.
